colorpicker/src/main/java/org/catrobat/paintroid/colorpicker/HSVColorPickerView.kt

```diff
@@ -0,0 +1,434 @@
+/*
+ * Paintroid: An image manipulation application for Android.
+ * Copyright (C) 2010-2021 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.paintroid.colorpicker
+
+import android.content.Context
+import android.graphics.BitmapFactory
+import android.graphics.BitmapShader
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.LinearGradient
+import android.graphics.Paint
+import android.graphics.Point
+import android.graphics.PorterDuff
+import android.graphics.PorterDuffXfermode
+import android.graphics.RectF
+import android.graphics.Shader
+import android.util.AttributeSet
+import android.view.MotionEvent
+import android.view.View
+import kotlin.math.max
+import kotlin.math.min
+
+private const val BORDER_WIDTH_PX = 1f
+private const val ALPHA_PANEL_HEIGHT = 21f
+private const val HUE_PANEL_WIDTH = 30f
+private const val PANEL_SPACING = 10f
+private const val PALETTE_CIRCLE_TRACKING_RADIUS = 5f
+private const val RECTANGLE_TRACKER_OFFSET = 2f
+private const val HUE = 360f
+private const val CONSTANT_1 = 3
+private const val CONSTANT_2 = 0xff
+
+class HSVColorPickerView : View {
+    private var alphaPanelHeight = ALPHA_PANEL_HEIGHT
+    private var panelSpacing = PANEL_SPACING
+    private var paletteCircleTrackerRadius = PALETTE_CIRCLE_TRACKING_RADIUS
+    private var rectangleTrackerOffset = RECTANGLE_TRACKER_OFFSET
+    private var huePanelWidth = HUE_PANEL_WIDTH
+    private var density = 1f
+    private var alpha = CONSTANT_2
+    private var hue = HUE
+    private var sat = 0f
+    private var brightness = 0f
+    private val sliderTrackerColor = -0xa0a0b
+    private val borderColor = -0x919192
+    private var drawingOffset = 0f
+    private var satValPaint: Paint
+    private var satValTrackerPaint: Paint
+    private var huePaint: Paint
+    private var hueTrackerPaint: Paint
+    private var alphaPaint: Paint
+    private var borderPaint: Paint
+    private var checkeredPaint: Paint
+    private var onColorChangedListener: OnColorChangedListener? = null
+    private var drawingRect: RectF? = null
```

Comment:
all the rects get initialized in the init block, and they never get set to null aftwards, so I think they can be a non nullable val

Reply:
yeah, nice catch ;D